SILGen: Add experimental TSan instrumentation for inout accesses.

devincoughlin opened the pull request:
Add SILGen instrumentation to treat inout accesses as Thread Sanitizer writes.
The goal is to catch races on inout accesses even when there is a not an
llvm-level read/write to a particular address. Ultimately
this will enable TSan to, for example, report racy writes to distinct
stored properties of a common struct as a data race.

This instrumentation is off by default. It can be enabled with the
'enable-experimental-tsan-inout-instrumentation' frontend flag.

The high-level approach is to add a SIL-level builtin that represents a call
to a TSan routine in compiler-rt. Then, when emitting an address for an LValue
as part of an inout expression, we call this builtin for each path component
that represents an LValue. I've added an 'isRValue()' method to PathComponent
that tracks whether a component represents an RValue or an LValue. Right the
only PathComponent that sometimes returns 'true' is ValueComponent().

For now, we're instrumenting only InoutExprs, but in the future it probably
makes sense to instrument all LValue accesses. In this patch I've
added a 'TSanKind' parameter to SILGenFunction::emitAddressOfLValue() and
its helpers to limit instrumentation to inout accesses. I envision that this
parameter will eventually go away.

jckarter commented on include/swift/Runtime/RuntimeFunctions.def:
```diff
@@ -1137,6 +1137,13 @@ FUNCTION(GetErrorValue, swift_getErrorValue, C_CC,
+FUNCTION(TSanInoutAccess, __tsan_write8, C_CC,
+         RETURNS(VoidTy),
+         ARGS(Int8PtrTy),
+         ATTRS(NoUnwind))
```
Is the plan to eventually use an appropriately sized `__tsan_writeN` call? Is there a dynamically-sized one for generic code?

devincoughlin replied:
That is the long-term plan. There is not a dynamically-sized callback for callback, so we (Kuba) will have to add one to compiler-rt. That is lower priority since we can get most of the benefits of checking without it. Instead we'll use __tsan_write1 (rather than 8, as I have here) for all inout accesses since that will avoid false positives from not including the size at the cost of false negatives.

jckarter commented on lib/SILOptimizer/Mandatory/DIMemoryUseCollector.cpp:
```diff
@@ -835,6 +850,11 @@ void ElementUseCollector::collectUses(SILValue Pointer, unsigned BaseEltNo) {
+    // Sanitizer instrumentation is not user visible, so it should not
+    // count as a use and must not affect compile-time diagnostics.
+    if (isSanitizerInstrumentation(User, Module.getASTContext()))
+      continue;
```
We can probably remove any `tsanInoutAccess` operations that happen prior to a storage slot's definite initialization, since the DI analysis should prevent escaping and ensure races are impossible before the initialization occurs.

devincoughlin replied:
Where would you recommend removing these operations? As a SIL pass after DI?

gottesmm replied:
@devincoughlin I would mess with that later TBH. Lets get this in as a first iteration.

jckarter commented:
Looks good to me so far.

gottesmm commented on lib/SIL/SILValue.cpp:
```diff
@@ -729,6 +729,9 @@ CONSTANT_OWNERSHIP_BUILTIN(Trivial, Assign)
 CONSTANT_OWNERSHIP_BUILTIN(Trivial, AtomicStore)
 CONSTANT_OWNERSHIP_BUILTIN(Trivial, Once)
+
+CONSTANT_OWNERSHIP_BUILTIN(Trivial, TSanInoutAccess)
```
Is there any reason why you added spacing here?

devincoughlin replied:
I expect to eventually have additional builtins for other kinds of TSan instrumentation (actual reads and writes) and thought I would group those together. It is premature at this point, though.

gottesmm commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -2635,7 +2635,8 @@ static void beginInOutFormalAccesses(SILGenFunction &gen,
       ManagedValue address = gen.emitAddressOfLValue(loc, std::move(inoutArg),
-                                                     AccessKind::ReadWrite);
+                                                     AccessKind::ReadWrite,
+                                                     /*forInout=*/true);
```
Can you use an enum instead? I.e. something like IsInitialization?

devincoughlin replied:
Will do.

gottesmm reviewed (commented):
Looks reasonable. Can we have some sil -> ll ir tests? All of your tests are swift only.

devincoughlin commented:
Thanks!

Yes, I will add IRGen tests.

devincoughlin commented:
- Updated to change the boolean 'forInout' parameter to an enum.
- At @kubamracek 's request, added a test for instrumented computed class properties.
- The builtins and IRGen portions of this patch were separated out into to https://github.com/apple/swift/pull/7925

devincoughlin commented:
@swift-ci Please test

devincoughlin committed:
SILGen: Add experimental TSan instrumentation for inout accesses.

Add SILGen instrumentation to treat inout accesses as Thread Sanitizer writes.
The goal is to catch races on inout accesses even when there is a not an
llvm-level read/write to a particular address. Ultimately
this will enable TSan to, for example, report racy writes to distinct
stored properties of a common struct as a data race.

This instrumentation is off by default. It can be enabled with the
'enable-experimental-tsan-inout-instrumentation' frontend flag.

The high-level approach is to add a SIL-level builtin that represents a call
to a TSan routine in compiler-rt. Then, when emitting an address for an LValue
as part of an inout expression, we call this builtin for each path component
that represents an LValue. I've added an 'isRValue()' method to PathComponent
that tracks whether a component represents an RValue or an LValue. Right the
only PathComponent that sometimes returns 'true' is ValueComponent().

For now, we're instrumenting only InoutExprs, but in the future it probably
makes sense to instrument all LValue accesses. In this patch I've
added a 'TSanKind' parameter to SILGenFunction::emitAddressOfLValue() and
its helpers to limit instrumentation to inout accesses. I envision that this
parameter will eventually go away.

devincoughlin commented:
Updated to fix a bad rebase:
- Use 'TSanKind::InoutAccess' in beginInOutFormalAccesses()
- Pass -sanitize=thread in SILGen test

Also XFAIL on Linux (where TSan is not supported).

devincoughlin commented:
@swift-ci Please test

devincoughlin commented:
Add a // REQUIRES: tsan_runtime to the SILGen test so we don't try to run it when TSan is not available (such as on 32-bit simulator).

devincoughlin commented:
@swift-ci Please test

devincoughlin commented:
@swift-ci Please smoke test and merge

gparker42 commented:
Test tsan-inout failed in CI. I'm reverting in #8162 unless you can fix it first.

gparker42 commented:
(Failure was swift-incremental-RA build 13266 in Apple CI. I don't see any failures in swift.org CI.)
